pubtools/_quay/security_manifest_pusher.py

```diff
@@ -0,0 +1,535 @@
+        ]
+
+    def get_multiarch_item_security_manifests(
+        self, item: object, dir_path: str
+    ) -> list[DigestSecurityManifest]:
+        """
+        Get security manifests of a multiarch image.
+
+        Each architecture has its own security manifest. Older images don't have security manifests,
+        so failure to extract them doesn't result in an error.
+
+        Args:
+            item (ContainerPushItem):
+                Push item of a multiarch image.
+            dir_path (str):
+                Directory path where the manifest files should be created.
+        Returns [DigestSecurityManifest]:
+            Extracted security manifests.
+        """
+        image_manifests = []
+        source_ref = item.metadata["pull_url"]
+        # Get only repo so that we can attach arch digests to it
+        source_repo = source_ref.split(":")[0]
+        manifest_list = self.src_quay_client.get_manifest(
+            source_ref, media_type=QuayClient.MANIFEST_LIST_TYPE
+        )
+
+        for arch in manifest_list["manifests"]:
+            image_ref = f"{source_repo}@{arch['digest']}"
+            security_manifest_path = os.path.join(
+                dir_path,
+                f"security_manifest_{arch['platform']['architecture']}_{uuid.uuid4().hex}.json",
+            )
+
+            result = self.cosign_get_security_manifest(image_ref, security_manifest_path)
+            if not result:
+                LOG.warning(
+                    f"Image {source_ref} with architecture {arch['platform']['architecture']}"
+                    " doesn't contain a security manifest"
+                )
+            else:
+                image_manifests.append(
+                    DigestSecurityManifest(
+                        digest=arch["digest"], security_manifest_path=security_manifest_path
+                    )
+                )
+
+        # I wonder if this should be treated as an error
+        if image_manifests and len(image_manifests) != len(manifest_list["manifests"]):
+            LOG.error(
+                f"Only some architectures of multiarch image {source_ref} have a security manifest"
+            )
+
+        return image_manifests
+
+    def push_item_security_manifests(self, item: object):
+        """
+        Generate and push container security manifest images of a push item.
+
+        Args:
+            item (ContainerPushItem):
+                Push item.
+        """
+        is_source_image = (
+            item.metadata["build"].get("extra", {}).get("image", {}).get("sources_for_nvr", False)
+        )
+        try:
+            self.src_quay_client.get_manifest(
+                item.metadata["pull_url"], media_type=QuayClient.MANIFEST_LIST_TYPE
+            )
+        except Exception:
+            is_multiach_image = False
+        else:
+            is_multiach_image = True
+
+        with tempfile.TemporaryDirectory(prefix="security_manifest_") as tmp_dir:
+            if is_source_image:
```

Comment:
This logic is analogous to ContainerImagePusher's push_container_images method. To my knowledge, security manifests can only be present in multiarch and (possibly) V2S2 source images. Could there be other image types?

I think it's dangerous to write code where we don't even understand what type of data we might receive. So I think it's better to explicitly declare what type of images we support.

Reply:
I think security manifest can be present in multiarch or v2sch2 regardless of if it's source container or not. Note that only difference between copy_source_push_item and copy_v1_push_item is log message.

Reply:
To my knowledge, new push items can only be multiarch or source images. And old items don't contain security manifests because they only started to be generated recently. I can't think of a scenario where other image types may contain security manifests. If new image types are added, I imagine this change will have to be reflected in our implementation as well. Do you disagree with this approach?

Reply:
My point is, there's no extra logic needed to handle source containers compared to other v2sch2 containers, so there's no reason to restrict the code to push only source containers.

Reply:
In my opinion, we should have an understanding what data we'll be working with and our code should reflect our expectations of that data. If we write our code in a way that "sort of maybe" supports other types of data, I think we're setting ourselves up for failure. Besides, I've written the code to be easily extensible if other push item types need to be supported.

If you still insist that this method should support "other" v2s2 push items then I will change it, but I still don't think it's a good idea.

Thoughts?

Reply:
I don't understand your point. Build team doesn't ship v2s2 containers at all so how can they decide to ship secure manifests for them? 

If you check ContainerImagePusher, the code doesn't expect v2s2 to be shipped at all. Due to poorly written condition, the method "copy_v1_push_item" will be called if non-source v2s2 is encountered, which is obviously wrong. The rest of the code base doesn't support normal v2s2, so I argue that security_manifest_pusher shouldn't either.